src/backend/distributed/connection/placement_connection.c

```diff
@@ -229,272 +222,373 @@ GetPlacementConnection(uint32 flags, ShardPlacement *placement, const char *user
-			/* we'd not see the other connection's contents */
+			/*
+			 * If a placement has been read over multiple (uncommitted) connections
+			 * then a DDL command on a co-located placement may create a self-deadlock
+			 * if there exist some relationship between the co-located placements
+			 * (e.g. foreign key, partitioning).
+			 */
 			ereport(ERROR,
 					(errcode(ERRCODE_ACTIVE_SQL_TRANSACTION),
-					 errmsg("cannot establish new placement connection when DML has been "
-							"executed on existing placement connection")));
+					 errmsg("cannot perform DDL on a placement if a co-located placement "
+							"has been read over multiple connections")));
 		}
-		else
+		else if (foundModifyingConnection)
 		{
-			/* modifying xact should have executed DML/DDL */
-			Assert(false);
-		}
-	}
+			/* we already found a connection that performed writes and must use it */
 
-	/*
-	 * Search existing connections for a reusable connection.
-	 */
-	dlist_foreach(it, &placementEntry->connectionReferences)
-	{
-		ConnectionReference *connectionReference =
-			dlist_container(ConnectionReference, placementNode, it.cur);
+			if ((placementConnection->hadDDL || placementConnection->hadDML) &&
+				placementConnection->connection != connection)
+			{
+				/*
+				 * The current placement may have been modified over a different
+				 * connection. Neither connection is guaranteed to see all uncomitted
+				 * writes and therefore we cannot proceed.
+				 */
+				ereport(ERROR,
+						(errcode(ERRCODE_ACTIVE_SQL_TRANSACTION),
+						 errmsg("cannot perform query with placements that were "
+								"modified over multiple connections")));
+			}
+		}
+		else if (CanUseExistingConnection(flags, userName, placementConnection))
+		{
+			/* use the connection assigned to the current placement */
+			connection = placementConnection->connection;
 
-		if (CanUseExistingConnection(flags, userName, connectionReference))
+			if (placementConnection->hadDDL || placementConnection->hadDML)
+			{
+				/* this connection performed writes, we must use it */
+				foundModifyingConnection = true;
+			}
+		}
+		else if (placementConnection->hadDDL)
 		{
-			return connectionReference;
+			ereport(ERROR,
+					(errcode(ERRCODE_ACTIVE_SQL_TRANSACTION),
+					 errmsg("cannot establish new placement connection when DDL has "
+							"been executed on existing placement connection")));
 		}
-
-		/*
-		 * If not using the connection, verify that FOR_DML/DDL flags are
-		 * compatible.
-		 */
-		if (flags & FOR_DDL)
+		else if (placementConnection->hadDML)
 		{
-			/* would deadlock otherwise */
 			ereport(ERROR,
 					(errcode(ERRCODE_ACTIVE_SQL_TRANSACTION),
-					 errmsg("cannot establish new placement connection for DDL when "
-							"other connections exist")));
+					 errmsg("cannot establish new placement connection when DML has "
+							"been executed on existing placement connection")));
 		}
-		else if (flags & FOR_DML)
+		else if (accessType == PLACEMENT_ACCESS_DDL)
```

Comment:
At this point, there can only have been reads to co-located shard placements (otherwise `hadDML` would have been true), and DML does not conflict with reads.